### PR TITLE
Add support for Other Receipts to Intacct api model

### PR DESCRIPTION
**Pull request by guymaness2**
##### What does the change add or fix?

Enhanced KFS's Intacct api interface to support creation of Other Receipts in Intacct account.

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

* Enhanced KFS's Intacct api interface to support creation of Other Receipts in Intacct account.

---------

##### Who reported, requested, or paid for the change?

River Point Church

---------

##### Does this update or add options to the block UI?

See RockBlocks PR for details.

---------

##### What files does it affect?

* rocks.kfs.Intacct/Enums/PaymentMethod.cs
* rocks.kfs.Intacct/IntacctCheckingAccountList.cs
* rocks.kfs.Intacct/IntacctEndpoint.cs
* rocks.kfs.Intacct/IntacctModels.cs
* rocks.kfs.Intacct/IntacctOtherReceipt.cs
* rocks.kfs.Intacct/rocks.kfs.Intacct.csproj

---------

##### Is it a breaking change for other versions/clients?

none

### CODE REVIEW

File: rocks.kfs.Intacct/Enums/PaymentMethod.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+// limitations under the License.
+// </copyright>
+//
+using System.ComponentModel;
+
+namespace rocks.kfs.Intacct.Enums
+{
+    public enum PaymentMethod
+    {
+        [Description( "Printed Check" )]
```
Comment (guymaness2) [inline rocks.kfs.Intacct/Enums/PaymentMethod.cs:23]: Add "Name" decorator and use Intacct UI nomenclature to use in Block.

File: rocks.kfs.Intacct/IntacctOtherReceipt.cs [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+                        var dimension = rockKey.Split( '.' ).Last();
+                        customDimensionValues.Add( dimension, account.GetAttributeValue( rockKey ) );
+                    }
+                }
+
+                mergeFields.Add( "Account", account );
+                mergeFields.Add( "Batch", financialBatch );
+                mergeFields.Add( "Registrations", registrationLinks );
+                mergeFields.Add( "GroupMembers", groupMemberLinks );
+                mergeFields.Add( "CustomDimensions", customDimensionValues );
```
Comment (guymaness2) [inline rocks.kfs.Intacct/IntacctOtherReceipt.cs:358]: Add field for "Summary" and pass the entire bTran object.

File: rocks.kfs.Intacct/IntacctOtherReceipt.cs [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+
+                var receiptItem = new ReceiptLineItem
+                {
+                    GlAccountNo = glAccount,
+                    Amount = bTran.ProcessTransactionFees == 1 ? bTran.Amount - bTran.TransactionFeeAmount : bTran.Amount,
+                    Memo = DescriptionLava.ResolveMergeFields( mergeFields ),
+                    LocationId = locationId,
+                    DepartmentId = departmentId,
+                    ProjectId = bTran.Project,
+                    ClassId = classId
```
Comment (guymaness2) [inline rocks.kfs.Intacct/IntacctOtherReceipt.cs:379]: Need to set CustomFields property to customdimensions

File: rocks.kfs.Intacct/IntacctOtherReceipt.cs [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+                        ClassId = classId
+                    };
+                    otherReceipt.ReceiptItems.Add( feeLineItem );
+                }
+            }
+
+            return otherReceipt;
+        }
+
+        private List<string> GetCustomDimensions()
```
Comment (guymaness2) [inline rocks.kfs.Intacct/IntacctOtherReceipt.cs:402]: extract method for reuse

File: rocks.kfs.Intacct/Utils/TransactionHelpers.cs [hunk 5]
```diff
@@ -0,0 +1,214 @@
+﻿// <copyright>
+// Copyright 2019 by Kingdom First Solutions
```
Comment (nateh777) [inline rocks.kfs.Intacct/Utils/TransactionHelpers.cs:2]: Update Copyright year

File: rocks.kfs.Intacct/IntacctJournal.cs [hunk 6]
```diff
@@ -204,140 +205,29 @@ private List<JournalEntryLine> GetGlEntries( FinancialBatch financialBatch, ref
...
             //
-            var customDimensions = GetCustomDimensions();
+            var customDimensions = TransactionHelpers.GetCustomDimensions();
             var batchSummary = new List<GLBatchTotals>();
             foreach ( var summary in batchTransactionsSummary )
             {
                 var account = new FinancialAccountService( rockContext ).Get( summary.FinancialAccountId );
                 var customDimensionValues = new Dictionary<string, dynamic>();
-                var mergeFields = new Dictionary<string, object>();
-                account.LoadAttributes();
```
Comment (nateh777) [inline rocks.kfs.Intacct/IntacctJournal.cs:318]: Attributes still need to be loaded on the account, even though it gets passed into a method that runs the loadattributes method I am not sure if it updates the object at this level with the attributes, and is a bit weird and not readable to rely on that anyway. I would rather have it here still.

File: rocks.kfs.Intacct/IntacctOtherReceipt.cs [hunk 7]
```diff
@@ -219,159 +220,38 @@ private OtherReceipt BuildOtherReceipt( FinancialBatch financialBatch, ref strin
...
-            var customDimensions = GetCustomDimensions();
+            var customDimensions = TransactionHelpers.GetCustomDimensions();
 
             // Create Receipt Item for each entry within a grouping
             foreach ( var bTran in receiptTransactions )
             {
                 var account = new FinancialAccountService( rockContext ).Get( bTran.FinancialAccountId );
                 var customDimensionValues = new Dictionary<string, dynamic>();
-                var mergeFields = new Dictionary<string, object>();
-                account.LoadAttributes();
```
Comment (nateh777) [inline rocks.kfs.Intacct/IntacctOtherReceipt.cs]: Attributes still need to be loaded on the account, even though it gets passed into a method that runs the loadattributes method I am not sure if it updates the object at this level with the attributes, and is a bit weird and not readable to rely on that anyway. I would rather have it here still.

File: rocks.kfs.Intacct/IntacctOtherReceipt.cs [hunk 8]
```diff
@@ -219,159 +220,38 @@ private OtherReceipt BuildOtherReceipt( FinancialBatch financialBatch, ref strin
...
-                    debugLava = mergeFields.lavaDebugInfo();
-                }
+                    Account = account,
+                    Batch = financialBatch,
+                    Registrations = registrationLinks,
+                    GroupMembers = groupMemberLinks,
+                    Summary = bTran,
+                    CustomDimensions = customDimensions
+                };
+                Dictionary<string, object> mergeFields = TransactionHelpers.GetMergeFieldsAndDimensions( ref debugLava, customDimensionValues, mergeFieldObjects );
```
Comment (nateh777) [inline rocks.kfs.Intacct/IntacctOtherReceipt.cs:246]: Did you test the lava debug? If not, can you? Just make sure it goes into the method and comes back out okay? I am expecting it should, but just want to be sure.

**Review by nateh777 — CHANGES_REQUESTED** [review]
@guymaness2 see comments on a few minor modifications. Also you tested both other receipts mode and regular export mode, correct? Just want to make sure.

File: rocks.kfs.Intacct/IntacctJournal.cs (hunk 6 above)
Comment (guymaness2) [reply rocks.kfs.Intacct/IntacctJournal.cs:318]: @nateh777  Do you mean remove it from the method and put it back here, or bring it back here and leave it in both places?

Comment (nateh777) [reply rocks.kfs.Intacct/IntacctJournal.cs:318]: I think just bring the loadattributes at this level, then I don't think you will need it at the lower level, I think it is just clearer that way and more guaranteed to work compared to I am not sure if the other way goes back up. Though you could leave it in both places I don't think it is necessary.

**Pull request merged**